proxy/proxy.go

```diff
@@ -242,6 +244,28 @@ func (p *Proxy) copyData(dst net.Conn, src net.Conn) {
 			p.Logger.Printf("error: %s", err)
 		}
 	}
+
+	switch srcConn := src.(type) {
+	case *net.TCPConn:
+		srcConn.SetLinger(0)
+		srcConn.CloseRead()
+	case *net.UnixConn:
+		srcConn.CloseRead()
+	default:
+		src.Close()
+	}
+
+	switch dstConn := dst.(type) {
+	case *net.TCPConn:
+		dstConn.SetLinger(0)
+		dstConn.CloseWrite()
+	case *tls.Conn:
+		dstConn.CloseWrite()
```

Comment:
Based on the docs for crypto/tls, this does not call CloseWrite for the underlying connection that the *tls.Conn is wrapping. Is that something we need to do here?

Reply:
I guess we can't really do that, but the fact that this is effectively non-symmetric to the above switch block and that it doesn't affect the underlying connection at all makes me think maybe we should call Close() here for TLS connections

Reply:
If i good understand the tls, there is no "close": if a server want to close the ssl connection, first sould send a message "i want to close the connection" (thats the close notify), and after the other side receives this notify, closes the socket, and server receive the other notfy, and close its socket too. https://security.stackexchange.com/questions/82028/ssl-tls-is-a-server-always-required-to-respond-to-a-close-notify
(Thats why, there is no closeread). 
The tls socket is not stay open, because the other copyData thread (which is reading the tls socket) will close the whole tls connection after the ssl notify.

i played with this and other "close solutions for proxies" found in github, but this was the best (actually the one, which is really worked good)

i tested with larget conn count (~10k paralell) and didnt got errors or the race conditions.

Reply:
Out of curiosity, how did you test? In client or server mode? I'd love to incorporate something like that into our integration test suite.

Reply:
in server mode: it was a ssl reverse proxy before a simple webserver.